src/profile-logic/marker-data.js

```diff
@@ -239,150 +238,283 @@ export function deriveMarkersFromRawMarkerTable(
-      // This is implemented by AutoProfilerTracing in Gecko.
-      //
-      // In this function we convert both of these raw markers into a single
-      // marker with a non-null duration.
-      //
-      // We also handle nested markers by assuming markers of the same type are
-      // never interwoven: given input markers startA, startB, endC, endD, we'll
-      // get 2 markers A-D and B-C.
-
-      const time = rawMarkers.time[i];
-      const nameStringIndex = rawMarkers.name[i];
-      if (data.interval === 'start') {
-        let markerBucket = openMarkers.get(nameStringIndex);
-        if (markerBucket === undefined) {
-          markerBucket = [];
-          openMarkers.set(nameStringIndex, markerBucket);
-        }
+      });
+      continue;
+    }
+
+    // Depending on the type we have to do some special handling.
+    switch (data.type) {
+      case 'tracing': {
+        // Markers are created from two distinct raw markers that are created at
+        // the start and end of whatever code that is running that we care about.
+        // This is implemented by AutoProfilerTracing in Gecko.
+        //
+        // In this function we convert both of these raw markers into a single
+        // marker with a non-null duration.
+        //
+        // We also handle nested markers by assuming markers of the same type are
+        // never interwoven: given input markers startA, startB, endC, endD, we'll
+        // get 2 markers A-D and B-C.
+        //
+        // It also happens that we don't have one side of the pair, in that case
+        // we still insert a marker and try to fill it with sensible values.
+        if (data.interval === 'start') {
+          let openMarkersForName = openTracingMarkers.get(name);
+          if (!openMarkersForName) {
+            openMarkersForName = [];
+            openTracingMarkers.set(name, openMarkersForName);
+          }
+          openMarkersForName.push(i);
 
-        markerBucket.push({
-          start: time,
-          name: stringTable.getString(nameStringIndex),
-          dur: 0,
-          title: null,
-          data,
-        });
-      } else if (data.interval === 'end') {
-        const markerBucket = openMarkers.get(nameStringIndex);
-        let marker;
-        if (markerBucket && markerBucket.length) {
-          // We already encountered a matching "start" marker for this "end".
-          marker = markerBucket.pop();
+          // We're not inserting anything to matchedMarkers yet. We wait for the
+          // end marker for that so that we know about the duration.
+          //
+          // We'll loop at all open markers after the main loop.
+        } else if (data.interval === 'end') {
+          const openMarkersForName = openTracingMarkers.get(name);
+          let startIndex;
+          if (openMarkersForName) {
+            startIndex = openMarkersForName.pop();
+          }
+          if (startIndex !== undefined) {
```

Comment:
Optional: I find the double negatives for an if check to be confusing. My preference is to switch the ordering to be `if (startIndex === undefined)`.

Reply:
Oh, and I believe there is a second `!== undefined` below as well.

Reply:
Mmm I see what you mean and that's usually my preference too. But in that case, this is actually the positive case, I mean that is the case where we actually found a startIndex, that's why I did it this way.
I'll look at the code again and see if this is better to reverse the order.

Reply:
I kept it like this. I see this condition `if (startIndex !== undefined)` as a better `if (startIndex)` (which obviously isn't right here).